Add information in the readme for spek tests (#146)
pusher/chatkit-android: README.md

credential from the [dashboard](https://dash.pusher.com/chatkit/).

*Important:* The tests will delete any and all resources associated with
the instance you provide. Create a new instance for testing purposes, and do not
share it anywhere else.

To run the tests directly from Android Studio you will need to install the
[Spek](https://plugins.jetbrains.com/plugin/8564-spek/) plugin, and
[Spek Framework](https://plugins.jetbrains.com/plugin/10915-spek-framework/).
Both plugins are needed as older tests are written using Spek 1, and newer
tests are using Spek 2.
To do this go to Android Studio -> Preferences -> Plugins ->
search for Spek, and Spek Framework and install both plugins. If you open a Spek test file
(e.g. RoomSpek), you should now see green play buttons to run each test
(or a test group from a file).

You will need to add your Chatkit test instance credentials to the VM — to do this edit
the run configurations -> select Spek on the left -> in the VM options field
enter the following:

```
-Dchatkit_integration_locator=<INSTANCE_LOCATOR> -Dchatkit_integration_key=<SECRET_KEY>
```

To run the Gradle test task, you will need to add your Chatkit test instance

> credentials to your global `gradle.properties` file:

Reviewer: ```suggestion
credentials to your global `~/.gradle/gradle.properties` file:
```

## Testing

The SDK has integration tests which run against a real Chatkit server.

> Firstly you will need to install the

Reviewer: ```suggestion
The tests can be run using Gradle `test` task.

To run the tests directly from IDE you will need to install the
```

> [Spek](https://www.spekframework.org/setup-android/) plugin, and Spek Framework in

Reviewer: ```suggestion
[Spek](https://plugins.jetbrains.com/plugin/8564-spek/) plugin, and
[Spek Framework](https://plugins.jetbrains.com/plugin/10915-spek-framework/) in
```

> Android Studio. To do this go to Android Studio -> Preferences -> Plugins ->

Reviewer: ```suggestion
Android Studio (both plugins are needed as old tests are written using Spek 1,
whilst the new/changed ones are using Spek 2).
To install the plugins go to Android Studio -> Preferences -> Plugins ->
```

> search for Spek, and Spek Framework and install. If you open a Spek test file, you

Reviewer: ```suggestion
search for Spek and install the both plugins. If you open a Spek test file, you
```

> should now see green play buttons to run each test (or the file).

Reviewer: ```suggestion
should now see green play buttons to run each test (or a test group from a file).
```

In order for the tests to pass you must provide a Chatkit instance

> credential to the VM - to do this edit the run configurations -> select Spek

Reviewer: ```suggestion
credential to the VM — to do this edit the run configurations -> select Spek
```
dash as opposed to minus

Author: I don't see a Gradle option when I edit my run configuation? Could you show me where you see this so I can ensure what we're telling people to do makes sense? If it helps, I don't have any run configuration enabled for Gradle, I run `gradle test` using the gradle menu.

on the left -> in the VM options field enter the following:

```

> -Dchatkit_integration_locator=xxxx -Dchatkit_integration_key=yyy

Reviewer 2: Including the `-D` is helpful, but were the value placeholders not clearer before?

Author: We decided that having the gradle.properties never worked for anyone, so let's keep with what worked for us. 

I figured writing `-Dchatkit_integration_locator=<instance locator> -Dchatkit_integration_key=<integration key>` said the same thing twice - having some characters without a space character in them makes it easier for people to replace?

Reviewer: Maybe I'm paying too much attention to detail but wouldn't it be nice to have equal number of x's and y's? ;)

Reviewer 2: `-Dchatkit_integration_key=<instance key>` not `<integration key>`, which is where I think it clarifies.

Reviewer: ```suggestion
-De2e_tests_chatkit_instance_locator=xxx -De2e_tests_chatkit_secret_key=yyy
```
to make things more clear (referencing how they are called in the dashboard)? also it may be worth adding a prefix explicitly conveying information what is it for (e2e tests)

Reviewer 2: (Would need further changes to consume them from those new names)

Reviewer: Of course, just two lines in a single [file](https://github.com/pusher/chatkit-android/blob/bbd6b0402ac4d773667453e2239c9a10a8f502c2/chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/InstanceTestFixtures.kt#L9).

Author: Have updated it to read `-Dchatkit_integration_locator=<INSTANCE_LOCATOR> -Dchatkit_integration_key=<SECRET_KEY>` which matches the words we use in the dashboard.

## Testing

The SDK has integration tests which run against a real Chatkit server.

The tests can be run using the Gradle `test` task with the VM options mentioned below set.
To run the tests directly from Android Studio you will need to install the
[Spek](https://plugins.jetbrains.com/plugin/8564-spek/) plugin, and
[Spek Framework](https://plugins.jetbrains.com/plugin/10915-spek-framework/).

> Both plugins are needed as older tests are written using Spek1, and newer

Reviewer: The authors call it: Spek 1 (with space), precisely they write Spek `1.x` and Spek `2.x`. both on top of their [page](https://www.spekframework.org/) and [GH](https://github.com/spekframework/spek/#current-state-of-the-project). I think we should follow the same naming (with space so that Spek is a word), just skipping the monospaced formatting and `.x` for ver no as that's more typographical/special style.

tests are using Spek2.
To do this go to Android Studio -> Preferences -> Plugins ->
search for Spek, and Spek Framework and install both plugins. If you open a Spek test file
(e.g. RoomSpek), you should now see green play buttons to run each test
(or a test group from a file).

> 

Reviewer: redundant empty line

In order for the tests to pass you must provide a Chatkit instance
credential from the [dashboard](https://dash.pusher.com/chatkit/)

> to the VM - to do this edit the run configurations -> select Spek

Reviewer: I had a suggestion here about adding "or Gradle" and using em dash instead of minus which is correct from typographical point of view. Indeed there's a reference to running via Gradle at the top and to invoke Gradle from IDE those two system properties need to be passed to the JVM running it.

Author: What do you mean by adding "or Gradle" sorry? I've updated the minus to be an emdash as per the symbol from https://www.thepunctuationguide.com/em-dash.html

Reviewer: I mean "select Spek or Gradle", as I suggested previously [here](https://github.com/pusher/chatkit-android/pull/146#discussion_r370222783). Depending if a user wants to run tests directly from the IDE (which is useful for running individual tests) or via Gradle, they will need to set/update their corresponding run configuration. Setting configuration just for Spek won't work for Gradle.

Author: We talked through this yesterday in person - I've added a section explaining to add your Chatkit test instance credentials into the global gradle.properties file.

chatkit_integration_key=<instance key>
```

> ### Spek Tests

Reviewer: I wouldn't create a subsection if there's only one subsection. It could be just a short paragraph about running from the IDE, something like that: "To run the tests from the IDE you need to instal Spek and Spek Framework plugins. Both plug-ins are needed as some of the tests are written using Spek 1 and some (new ones) are written using Spek 2".

Author: How do people install Spek? through gradle dependency? And Spek2 is the plugin? I figured we needed both the plugin + dependency before we used Spek2?

Reviewer: dependency is already declared in build.gradle — everything works via gradle without the need to install anyting. if the users want to run tests directly from the IDE (not through gradle) then they need to install the IDE plugins

Our tests are currently written using the [Spek Framework](https://www.spekframework.org/setup-android/).
And you will need to install the Spek Plugin in Android Studio to be able to run them.

> You may find that you will also need to edit the run configuration and add your

Reviewer: I would make it explicit as it's not "you may find" but "you will definitely need". That point is also already covered above. I would just make it more explicit like it is below and remove `local.properties` stuff as I don't think that ever worked as that file is just read by Android Gradle Plugin (not Gradle).

Author: Gotcha! I didn't want to change too much what was already there in case it worked for others. But you're probably right, the other one probably didn't work for anyone else :)

chatkit locator and chatkit key like this:

```
-Dchatkit_integration_locator=xxxx -Dchatkit_integration_key=yyy
```

> Happy testing!

Reviewer: I would skip it as that may be interpreted like we expect our users to do testing of our SDK, which is not the case. It's more if someone wants to contribute and use SDK's tests to verify their contribution.

Author: Sounds fair! Have removed.
